linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -1713,6 +1716,279 @@ private[optimizer] abstract class OptimizerCore(
     case _                 => false
   }
 
+  /** Tries to insert `valTree` in place of the (unique) occurrence of `valName` in `body`.
+   *
+   *  This function assumes that `valName` is used only once, and only inside
+   *  `body`. It does not assume that `valTree` or `body` are side-effect-free.
+   *
+   *  The replacement is done only if we can show that it will not affect
+   *  evaluation order. In practice, this means that we only replace if we find
+   *  the occurrence of `valName` in the first evaluation context of `body`.
+   *  In other words, we verify that all the expressions that will evaluate
+   *  before `valName` in `body` are pure.
+   *
+   *  We consider a `VarRef(y)` pure if `valTree` does not contain any
+   *  assignment to `y`.
+   *
+   *  For example, we can replace `x` in the following bodies:
+   *
+   *  {{{
+   *  x
+   *  x + e
+   *  x.foo(...)
+   *  x.f
+   *  e + x                 // if `e` is pure
+   *  e0.foo(...e1, x, ...) // if `e0` is pure and non-null, and the `...e1`s are pure
+   *  if (x) { ... } else { ... }
+   *  }}}
+   *
+   *  Why is this interesting? Mostly because of inlining.
+   *
+   *  Inlining tends to create many bindings for the receivers and arguments of
+   *  methods. We must do that to preserve evaluation order, and not to evaluate
+   *  them multiple times. However, very often, the receiver and arguments are
+   *  used exactly once in the inlined body, and in-order. Using this strategy,
+   *  we can take the right-hand-sides of the synthetic bindings and inline them
+   *  directly inside the body.
+   *
+   *  This in turn allows more trees to remain JS-level expressions, which means
+   *  that `FunctionEmitter` has to `unnest` less often, further reducing the
+   *  amount of temporary variables.
+   *
+   *  ---
+   *
+   *  Note that we can never cross any potential undefined behavior, even when
+   *  the corresponding semantics are `Unchecked`. That is because the
+   *  `valTree` could throw itself, preventing the normal behavior of the code
+   *  to reach the undefined behavior in the first place. Consider for example:
+   *
+   *  {{{
+   *  val x: Foo = ??? // maybe null
+   *  val y: Int = if (x == null) throw new Exception() else 1
+   *  x.foo(y)
+   *  }}}
```

Comment:
This is a good example, but it took me a while to understand the point it is making. (It was not clear to me, what we are inlining).

Maybe add the following sentence after the codeblock:

> We cannot inline `y` in this example, because that would change observable behavior if `x` is null.

Reply:
Ah yes, indeed, that was not clear. I added the suggested sentence. I also replace the `???` by `...` in the definition of `x`. the `???` suggested too strongly that it would throw itself.